runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/auth/HttpSigner.kt

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.auth
+
+import aws.smithy.kotlin.runtime.client.ExecutionContext
+import aws.smithy.kotlin.runtime.http.request.HttpRequestBuilder
+import aws.smithy.kotlin.runtime.util.InternalApi
+
+/**
+ * Represents a component capable of signing an HTTP request
+ */
+@InternalApi
+public interface HttpSigner {
+    public companion object {
+        /**
+         * A no-op signer that does nothing with the request
+         */
+        public val NONE: HttpSigner = object : HttpSigner {
+            override suspend fun sign(context: ExecutionContext, request: HttpRequestBuilder) { }
+        }
+    }
+
+    /**
+     * Sign the provided HTTP request (e.g. add AWS SigV4 headers, Bearer token header, etc)
+     */
+    public suspend fun sign(context: ExecutionContext, request: HttpRequestBuilder)
```

Comment:
**Question**: We generally have to make a copy of a request before signing it since signing happens _per-attempt_ and thus we can't modify the original. Since we're adding this interface and revising `AwsSigningMiddleware` I have to wonder: does the copying belong in the `HttpSigner` interface?

```interface
interface HttpSigner {
    // May return a signed copy or, in rare cases, the request arg unmodified (a no-op signature)
    suspend fun sign(context: ExecutionContext, request: HttpRequest): HttpRequest
}
```